### DISCUSSION TITLE
Load 'load' function only once

### DISCUSSION THREAD

**Discussion by jdgamble555** [sveltejs/kit#7816] [Q&A]
How can I load the `load` function only once when the app is first loaded? I have a route with a `[slug]` that I need to run on the server on the first load, but would like to not reload the load function on the second loads to the same component (different path).

I want to load the load function once, then just reload my component without the load function.

J

**Comment by david-plugge** [comment]
Can you give a code example?

**Comment by jdgamble555** [comment]
I'm just asking as a concept.  I really just need to load data from the server on the first load (for seo), then load in the browser on any route changes afterwards.  There isn't always a need to run from the server except for the first page load of the app.  This would include any route afterwards, or the same component with a different parameter etc.

Is this possible?

J

**Reply by teemingc** [reply]
If you use a layout load function in the root of /routes, it should intelligently rerun instead of on every page change.
https://kit.svelte.dev/docs/load#layout-data

**Reply by jdgamble555** [reply]
My url changes, so it does rerun. I would like control over this, so that won't help. Intelligently doesn't help if you can't control when it runs manually.

J

**Reply by teemingc** [reply]
Does this example help? I'm not too sure when the load will re-run (aside from client side invalidate, page reload, or other if statements) but in my experience the root layout doesn't re-run that often.
https://stackblitz.com/edit/sveltejs-kit-template-default-pcrtbr?file=src/routes/+layout.js

**Reply by jdgamble555** [reply]
So you're just caching the data. If you need to use the slug within the load function, sveltekit will automatically rerun (if url or params changes). 

I don't want to run the load function or cache it at all except for on the first load. It will contain data that depends on the slug.

Angular Universal, for example, runs a resolver ONLY on the first load. This is useful for SEO, but then you don't need to rerun the resolver on each route load, as SEO bots won't keep clicking. Your app would just do what it needs to do on the client side, or fetch some api.

___

Maybe this could help to think about it in another way.  I want to load something from the server ONLY ONCE on first app start for SEO purposes that is dependent on the route.  How can I do this?  Perhaps the load function is not the way to do this, as it loads every time a route changes.  Is there a way to load something from the server just once on app start besides the load function (or load function within `+server.ts`)?

J

**Reply by teemingc** [reply]
You should be able to achieve this with in an `onMount` function in a root `+layout.svelte`. That way it will only run once (the first time the layout is mounted) when the app starts on any page. Am I understanding this correctly?

Alternatively, setting a cookie in the root layout load function would help mark if a user has visited the app before. Then, the existence of the cookie could be checked in subsequent loads to prevent running the same code again.

**Reply by jdgamble555** [reply]
The `onMount()` runs client side, so wouldn't help with SEO. Your second option may be a workaround though.  I think there should be a option in the load function `loadOnce = true`... anyway, thanks for the help. :) I will just find a workaround, as it seems this isn't possible as is.

J

**Reply by teemingc** [reply]
> The `onMount()` runs client side, so wouldn't help with SEO. Your second option may be a workaround though.  I think there should be a option in the load function `loadOnce = true`... anyway, thanks for the help. :) I will just find a workaround, as it seems this isn't possible as is.
> 
> 
> 
> J

I agree. It would be useful to have the option for a load function to only run once. Perhaps an issue should be opened for it

**Reply by jdgamble555** [reply]
Done - https://github.com/sveltejs/kit/issues/7822
